tools/promql-generator/promql-generator.go

```diff
@@ -0,0 +1,354 @@
+	}
+
+	err = cfg.validate()
+	if err != nil {
+		printUsageAndExit(err)
+	}
+
+	if *cfg.labelsSource.flag == "file" {
+		series, err = labelSetsFromFile(*cfg.labelsFile.flag)
+	} else {
+		series, err = labelSetsFromPrometheus(*cfg.prometheusUrl.flag, *cfg.prometheusMatcher.flag)
+	}
+
+	if err != nil {
+		printUsageAndExit(err)
+	}
+
+	if err := generatePromQL(*cfg.promqlCount.config.flag, series); err != nil {
+		os.Exit(1)
+	}
+}
+
+// labelSetsFromPrometheus attempt to get label sets from Prometheus server - gets all label sets which matches the given matcher
+// address is the fully qualified URL of the Prometheus server
+// matcher is a string like {job=\"prometheus\"}
+// Note - the matcher will only find a label set used in the last 2 hours
+func labelSetsFromPrometheus(address string, matcher string) ([]model.LabelSet, error) {
+	client, err := api.NewClient(api.Config{
+		Address: address,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "create Prometheus client")
+	}
+	promAPI := v1.NewAPI(client)
+	ctx := context.Background()
+	now := time.Now()
+
+	labels, _, err := promAPI.Series(
+		ctx,
+		[]string{matcher},
+		now.Add(-2*time.Hour),
+		now,
+	)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "get label sets")
+	}
+
+	if len(labels) == 0 {
+		return nil, errors.New("no matching label sets found")
+	}
+
+	return labels, nil
+}
+
+// labelSetsFromFile constructs a []model.LabelSet from a json file defining the label sets
+func labelSetsFromFile(seriesFile string) ([]model.LabelSet, error) {
+	data, err := os.ReadFile(seriesFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []map[string]string
+	if err := json.Unmarshal(data, &result); err != nil {
+		return nil, err
+	}
+
+	if len(result) == 0 {
+		return nil, errors.New("no label sets found")
+	}
+
+	labels := make([]model.LabelSet, len(result))
+
+	for i, item := range result {
+		labelSet := make(model.LabelSet)
+		for k, v := range item {
+			labelSet[model.LabelName(k)] = model.LabelValue(v)
+		}
+		labels[i] = labelSet
+	}
+
+	return labels, nil
+}
+
+// generatePromQL generate the random PromQL queries for the given label sets and writes to stdout
+func generatePromQL(count int, labelSets []model.LabelSet) error {
+
+	now := time.Now()
+	rnd := rand.New(rand.NewSource(now.Unix()))
+	opts := []promqlsmith.Option{
+		promqlsmith.WithEnableOffset(true),
+		promqlsmith.WithEnableAtModifier(true),
+		promqlsmith.WithEnabledFunctions(availableFunctions()),
+		promqlsmith.WithEnabledBinOps(enabledBinops),
+		promqlsmith.WithEnableVectorMatching(true),
+	}
+	ps := promqlsmith.New(rnd, modelLabelSetToLabels(labelSets), opts...)
+
+	for range count {
+
+		expr := ps.WalkInstantQuery()
+		query := expr.Pretty(0)
+
+		fmt.Printf("%s\n", strings.ReplaceAll(query, "\n", ""))
+
+	}
+	return nil
+}
+
+// modelLabelSetToLabels converts []model.LabelSet to []labels.Labels
+func modelLabelSetToLabels(labelSets []model.LabelSet) []labels.Labels {
```

Comment:
Given we never use the `model.LabelSet` directly, should we parse the labels from the remote server or JSON file directly into `labels.Labels`? Then we wouldn't need to do this conversion.